File: apps/gateway/src/agent/tools/web.ts
// Web 工具：web_search（tavily/brave 分发）与 fetch_url（含 SSRF 防护）

import { type Tool, tool } from 'ai';
import { z } from 'zod';
import { decrypt } from '../../crypto';
import { type AgentSettingsRecord, getAgentSettings } from '../../db/agent';
import { wrapUntrusted } from './untrusted';

const WEB_SEARCH_MAX_RESULTS = 8;
const WEB_SEARCH_RESULT_MAX_BYTES = 8 * 1024;
const FETCH_URL_TIMEOUT_MS = 15_000;
const FETCH_URL_MAX_BODY_BYTES = 2 * 1024 * 1024;
const FETCH_URL_TEXT_MAX_BYTES = 16 * 1024;
const FETCH_URL_MAX_REDIRECTS = 3;

export interface WebSearchResultItem {
  title: string;
  url: string;
  snippet: string;
}

function truncateUtf8(text: string, maxBytes: number): string {
  const encoder = new TextEncoder();
  if (encoder.encode(text).length <= maxBytes) {
    return text;
  }
  let result = text;
  // 二分收敛到字节上限，避免逐字符循环
  let low = 0;
  let high = text.length;
  while (low < high) {
    const mid = Math.ceil((low + high) / 2);
    if (encoder.encode(text.slice(0, mid)).length <= maxBytes) {
      low = mid;
    } else {
      high = mid - 1;
    }
  }
  result = text.slice(0, low);
  return `${result}\n[truncated]`;
}

// ========== SSRF 防护 ==========

function normalizeHostname(hostname: string): string {
  let value = hostname.toLowerCase();
  if (value.startsWith('[') && value.endsWith(']')) {
    value = value.slice(1, -1);
  }
  return value;
}

// 规范 IPv4 八位组：0-255、十进制、无前导零
const CANONICAL_IPV4_OCTET = /^(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)$/;

function isCanonicalIpv4(host: string): boolean {
  const parts = host.split('.');
  return parts.length === 4 && parts.every((part) => CANONICAL_IPV4_OCTET.test(part));
}

// 数字形式 host（十进制/0x 十六进制/前导零八进制段，1-4 段）：
// 非规范点分四段时是 IPv4 字面量的混淆写法（如 2130706433、127.1、0177.0.0.1、0x7f000001）
function isNumericHost(host: string): boolean {
  const parts = host.split('.');
  if (parts.length === 0 || parts.length > 4) {
    return false;
  }
  return parts.every((part) => /^(0x[0-9a-f]+|\d+)$/.test(part));
}

export function isPrivateHostname(hostname: string): boolean {
  const host = normalizeHostname(hostname);

  if (host === 'localhost' || host.endsWith('.localhost')) {
    return true;
  }

  if (isCanonicalIpv4(host)) {
    const [a, b] = host.split('.').map(Number);
    if (a === 0 || a === 10 || a === 127) return true;
    if (a === 169 && b === 254) return true;
    if (a === 172 && b !== undefined && b >= 16 && b <= 31) return true;
    if (a === 192 && b === 168) return true;
    return false;
  }

  // 非规范数字形式（整数 IP、缺段、八进制/十六进制段）一律拒绝：
  // 正常网站不会用这种 host，放行只会留下绕过私网判断的口子
  if (isNumericHost(host)) {
    return true;
  }

  // IPv6（URL.hostname 对 IPv6 字面量返回带括号形式，已剥除）
  if (host.includes(':')) {
    if (host === '::1' || host === '::') return true;
    if (host.startsWith('fc') || host.startsWith('fd')) return true; // fc00::/7（含 fd00::/8）
    if (host.startsWith('fe80:')) return true; // link-local
    if (host.startsWith('::ffff:')) {
      return isPrivateHostname(host.slice('::ffff:'.length));
    }
  }

  return false;
}

function allowPrivateFetch(): boolean {
  return process.env.TMEX_AGENT_ALLOW_PRIVATE_FETCH === '1';
}

export function validateFetchUrl(rawUrl: string): { url: URL } | { error: string } {
  let url: URL;
  try {
    url = new URL(rawUrl);
  } catch {
    return { error: `Invalid URL: ${rawUrl}` };
  }
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    return { error: `Unsupported protocol: ${url.protocol} (only http/https are allowed)` };
  }
  if (!allowPrivateFetch() && isPrivateHostname(url.hostname)) {
    return { error: `Refusing to fetch private/internal address: ${url.hostname}` };
  }
  return { url };
}

// ========== web_search ==========

interface WebSearchDeps {
  fetchImpl?: typeof fetch;
  tavilyEndpoint?: string;
  braveEndpoint?: string;
}

async function searchTavily(
  apiKey: string,
  query: string,
  deps: Required<Pick<WebSearchDeps, 'fetchImpl' | 'tavilyEndpoint'>>
): Promise<WebSearchResultItem[]> {
  const response = await deps.fetchImpl(deps.tavilyEndpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      // Tavily 现行 API 使用 Bearer header，旧版接受 body api_key；两者都带以兼容
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      api_key: apiKey,
      query,
      max_results: WEB_SEARCH_MAX_RESULTS,
    }),
    signal: AbortSignal.timeout(FETCH_URL_TIMEOUT_MS),
  });
  if (!response.ok) {
    throw new Error(`Tavily search failed: HTTP ${response.status}`);
  }
  const payload = (await response.json()) as {
    results?: Array<{ title?: string; url?: string; content?: string }>;
  };
  return (payload.results ?? []).map((item) => ({
    title: item.title ?? '',
    url: item.url ?? '',
    snippet: item.content ?? '',
  }));
}

async function searchBrave(
  apiKey: string,
  query: string,
  deps: Required<Pick<WebSearchDeps, 'fetchImpl' | 'braveEndpoint'>>
): Promise<WebSearchResultItem[]> {
  const url = new URL(deps.braveEndpoint);
  url.searchParams.set('q', query);
  url.searchParams.set('count', String(WEB_SEARCH_MAX_RESULTS));
  const response = await deps.fetchImpl(url.toString(), {
    headers: {
      Accept: 'application/json',
      'X-Subscription-Token': apiKey,
    },
    signal: AbortSignal.timeout(FETCH_URL_TIMEOUT_MS),
  });
  if (!response.ok) {
    throw new Error(`Brave search failed: HTTP ${response.status}`);
  }
  const payload = (await response.json()) as {
    web?: { results?: Array<{ title?: string; url?: string; description?: string }> };
  };
  return (payload.web?.results ?? []).map((item) => ({
    title: item.title ?? '',
    url: item.url ?? '',
    snippet: item.description ?? '',
  }));
}

export interface CreateWebSearchToolOptions extends WebSearchDeps {
  settings?: AgentSettingsRecord;
}

/** searchProvider='none' 或未配置对应 key 时返回 null（不注册工具） */
export async function createWebSearchTool(
  options: CreateWebSearchToolOptions = {}
): Promise<Tool | null> {
  const settings = options.settings ?? getAgentSettings();
  const fetchImpl = options.fetchImpl ?? fetch;
  const tavilyEndpoint = options.tavilyEndpoint ?? 'https://api.tavily.com/search';
  const braveEndpoint = options.braveEndpoint ?? 'https://api.search.brave.com/res/v1/web/search';

  let search: ((query: string) => Promise<WebSearchResultItem[]>) | null = null;

  if (settings.searchProvider === 'tavily' && settings.tavilyApiKeyEnc) {
    const apiKey = await decrypt(settings.tavilyApiKeyEnc);
    search = (query) => searchTavily(apiKey, query, { fetchImpl, tavilyEndpoint });
  } else if (settings.searchProvider === 'brave' && settings.braveApiKeyEnc) {
    const apiKey = await decrypt(settings.braveApiKeyEnc);
    search = (query) => searchBrave(apiKey, query, { fetchImpl, braveEndpoint });
  }

  if (!search) {
    return null;
  }

  const searchFn = search;
  return tool({
    description: 'Search the web. Returns a JSON array of results with title, url and snippet.',
    inputSchema: z.object({
      query: z.string().min(1).describe('The search query.'),
    }),
    execute: async ({ query }) => {
      try {
        const results = await searchFn(query);
        return truncateUtf8(JSON.stringify(results), WEB_SEARCH_RESULT_MAX_BYTES);
      } catch (error) {
        return `Web search failed: ${error instanceof Error ? error.message : String(error)}`;
      }
    },
  });
}

// ========== fetch_url ==========

async function readBodyWithLimit(response: Response, maxBytes: number): Promise<string> {
  const reader = response.body?.getReader();
  if (!reader) {
    return '';
  }
  const chunks: Uint8Array[] = [];
  let total = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(value);
    total += value.length;
    if (total >= maxBytes) {
      await reader.cancel();
      break;
    }
  }
  const merged = new Uint8Array(Math.min(total, maxBytes));
  let offset = 0;
  for (const chunk of chunks) {
    const remaining = merged.length - offset;
    if (remaining <= 0) break;
    merged.set(chunk.subarray(0, Math.min(chunk.length, remaining)), offset);
    offset += Math.min(chunk.length, remaining);
  }
  return new TextDecoder('utf-8', { fatal: false }).decode(merged);
}

const STRIP_ELEMENT_SELECTORS = [
  'script',
  'style',
  'noscript',
  'template',
  'svg',
  'nav',
  'header',
  'footer',
  'aside',
  'iframe',
];

async function extractHtmlText(html: string): Promise<string> {
  let rewriter = new HTMLRewriter();
  for (const selector of STRIP_ELEMENT_SELECTORS) {
    rewriter = rewriter.on(selector, {
      element(element) {
        element.remove();
      },
    });
  }
  const cleaned = await rewriter.transform(new Response(html)).text();
  return cleaned
    .replace(/<[^>]*>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/&amp;/gi, '&')
    .replace(/&lt;/gi, '<')
    .replace(/&gt;/gi, '>')
    .replace(/&quot;/gi, '"')
    .replace(/&#39;/gi, "'")
    .replace(/[ \t\u00a0]+/g, ' ')
    .replace(/(?:[ \t]*\n[ \t]*){2,}/g, '\n\n')
    .trim();
}

export interface CreateFetchUrlToolOptions {
  fetchImpl?: typeof fetch;
}

export function createFetchUrlTool(options: CreateFetchUrlToolOptions = {}): Tool {
  const fetchImpl = options.fetchImpl ?? fetch;

  return tool({
    description:
      'Fetch a public http/https URL and return its readable text content (HTML is converted to plain text).',
    inputSchema: z.object({
      url: z.string().min(1).describe('The absolute http/https URL to fetch.'),
    }),
    execute: async ({ url: rawUrl }) => {
      let currentUrl = rawUrl;
      try {
        for (let redirects = 0; redirects <= FETCH_URL_MAX_REDIRECTS; redirects++) {
          const validated = validateFetchUrl(currentUrl);
          if ('error' in validated) {
            return validated.error;
          }

          const response = await fetchImpl(validated.url.toString(), {
            redirect: 'manual',
            signal: AbortSignal.timeout(FETCH_URL_TIMEOUT_MS),
            headers: { Accept: 'text/html,application/xhtml+xml,text/plain,*/*' },
          });

          if (response.status >= 300 && response.status < 400) {
            const location = response.headers.get('location');
            if (!location) {
              return `Fetch failed: HTTP ${response.status} redirect without Location header`;
            }
            currentUrl = new URL(location, validated.url).toString();
            continue;
          }

          if (!response.ok) {
            return `Fetch failed: HTTP ${response.status}`;
          }

          const contentType = response.headers.get('content-type') ?? '';
          const body = await readBodyWithLimit(response, FETCH_URL_MAX_BODY_BYTES);

          if (contentType.includes('text/html') || contentType.includes('application/xhtml')) {
            const text = await extractHtmlText(body);
            return wrapUntrusted(truncateUtf8(text, FETCH_URL_TEXT_MAX_BYTES), 'web');
          }

          return wrapUntrusted(truncateUtf8(body, FETCH_URL_TEXT_MAX_BYTES), 'web');
        }
        return `Fetch failed: too many redirects (>${FETCH_URL_MAX_REDIRECTS})`;
      } catch (error) {
        if (error instanceof DOMException && error.name === 'TimeoutError') {
          return `Fetch failed: timeout after ${FETCH_URL_TIMEOUT_MS}ms`;
        }
        return `Fetch failed: ${error instanceof Error ? error.message : String(error)}`;
      }
    },
  });
}

File: apps/gateway/src-rust/agent/tools/web.rs
use std::collections::BTreeMap;
use std::fmt;
use std::net::{IpAddr, Ipv4Addr};
use std::sync::{Arc, LazyLock};
use std::time::Duration;

use async_trait::async_trait;
use regex::Regex;
use reqwest::Url;

use super::{wrap_untrusted, UntrustedContentKind};
use crate::agent::{redact_secrets, AgentPortError};

pub const WEB_SEARCH_MAX_RESULTS: usize = 8;
pub const WEB_SEARCH_RESULT_MAX_BYTES: usize = 8 * 1024;
pub const FETCH_URL_TIMEOUT: Duration = Duration::from_secs(15);
pub const FETCH_URL_MAX_BODY_BYTES: usize = 2 * 1024 * 1024;
pub const FETCH_URL_TEXT_MAX_BYTES: usize = 16 * 1024;
pub const FETCH_URL_MAX_REDIRECTS: usize = 3;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WebSearchResultItem {
    pub title: String,
    pub url: String,
    pub snippet: String,
}

#[async_trait]
pub trait AgentWebSearch: Send + Sync {
    async fn is_configured(&self) -> Result<bool, AgentPortError> {
        Ok(true)
    }

    async fn search(
        &self,
        query: &str,
        max_results: usize,
    ) -> Result<Vec<WebSearchResultItem>, AgentPortError>;
}

#[async_trait]
pub trait AgentDnsResolver: Send + Sync {
    async fn resolve(&self, host: &str, port: u16) -> Result<Vec<IpAddr>, AgentPortError>;
}

#[derive(Clone, Copy, Debug, Default)]
pub struct TokioDnsResolver;

#[async_trait]
impl AgentDnsResolver for TokioDnsResolver {
    async fn resolve(&self, host: &str, port: u16) -> Result<Vec<IpAddr>, AgentPortError> {
        let addresses = tokio::net::lookup_host((host, port))
            .await
            .map_err(|error| AgentPortError::new(format!("DNS resolution failed: {error}")))?;
        let mut ips = addresses.map(|address| address.ip()).collect::<Vec<_>>();
        ips.sort();
        ips.dedup();
        Ok(ips)
    }
}

#[derive(Clone)]
pub struct ResolvedWebTarget {
    pub url: Url,
    pub resolved_ips: Vec<IpAddr>,
}

impl fmt::Debug for ResolvedWebTarget {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("ResolvedWebTarget")
            .field("scheme", &self.url.scheme())
            .field("host", &self.url.host_str())
            .field("port", &self.url.port_or_known_default())
            .field("resolved_ips", &self.resolved_ips)
            .finish()
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum WebHttpMethod {
    Get,
    Post,
}

#[derive(Clone)]
pub struct WebHttpRequest {
    pub method: WebHttpMethod,
    pub target: ResolvedWebTarget,
    pub timeout: Duration,
    pub max_body_bytes: usize,
    pub headers: BTreeMap<String, String>,
    pub body: Vec<u8>,
}

impl fmt::Debug for WebHttpRequest {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("WebHttpRequest")
            .field("method", &self.method)
            .field("target", &self.target)
            .field("timeout", &self.timeout)
            .field("max_body_bytes", &self.max_body_bytes)
            .field("header_names", &self.headers.keys().collect::<Vec<_>>())
            .field("body_len", &self.body.len())
            .finish()
    }
}

#[derive(Clone, PartialEq, Eq)]
pub struct WebHttpResponse {
    pub status: u16,
    pub headers: BTreeMap<String, String>,
    pub body: Vec<u8>,
}

impl fmt::Debug for WebHttpResponse {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("WebHttpResponse")
            .field("status", &self.status)
            .field("header_names", &self.headers.keys().collect::<Vec<_>>())
            .field("body_len", &self.body.len())
            .finish()
    }
}

#[async_trait]
pub trait AgentWebHttpTransport: Send + Sync {
    /// Implementations must disable automatic redirects and connect only to one of
    /// `request.target.resolved_ips`; resolving the hostname again would reopen DNS rebinding.
    async fn send(&self, request: WebHttpRequest) -> Result<WebHttpResponse, AgentPortError>;
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum FetchUrlPolicyError {
    #[error("invalid URL")]
    InvalidUrl,
    #[error("unsupported protocol: {0} (only http/https are allowed)")]
    UnsupportedProtocol(String),
    #[error("URLs containing credentials are not allowed")]
    EmbeddedCredentials,
    #[error("refusing to fetch private/internal address: {0}")]
    PrivateAddress(String),
    #[error("DNS resolution returned no addresses")]
    EmptyDnsResult,
    #[error(transparent)]
    Resolver(#[from] AgentPortError),
}

fn normalize_hostname(hostname: &str) -> String {
    hostname
        .trim_start_matches('[')
        .trim_end_matches(']')
        .to_ascii_lowercase()
}

fn is_canonical_ipv4(host: &str) -> bool {
    let parts = host.split('.').collect::<Vec<_>>();
    parts.len() == 4
        && parts.iter().all(|part| {
            !part.is_empty()
                && (part == &"0" || !part.starts_with('0'))
                && part.parse::<u8>().is_ok()
        })
}

fn is_numeric_host(host: &str) -> bool {
    let parts = host.split('.').collect::<Vec<_>>();
    !parts.is_empty()
        && parts.len() <= 4
        && parts.iter().all(|part| {
            let decimal = !part.is_empty() && part.bytes().all(|byte| byte.is_ascii_digit());
            let hexadecimal = part.strip_prefix("0x").is_some_and(|value| {
                !value.is_empty() && value.bytes().all(|b| b.is_ascii_hexdigit())
            });
            decimal || hexadecimal
        })
}

pub fn is_private_hostname(hostname: &str) -> bool {
    let host = normalize_hostname(hostname);
    if host == "localhost" || host.ends_with(".localhost") {
        return true;
    }
    if let Ok(ip) = host.parse::<IpAddr>() {
        return is_private_or_internal_ip(ip);
    }
    if is_canonical_ipv4(&host) {
        return host
            .parse::<Ipv4Addr>()
            .map(IpAddr::V4)
            .is_ok_and(is_private_or_internal_ip);
    }
    is_numeric_host(&host)
}

pub fn is_private_or_internal_ip(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(ip) => {
            let [a, b, ..] = ip.octets();
            a == 0
                || a == 10
                || a == 127
                || (a == 100 && (64..=127).contains(&b))
                || (a == 169 && b == 254)
                || (a == 172 && (16..=31).contains(&b))
                || (a == 192 && b == 168)
                || a >= 224
        }
        IpAddr::V6(ip) => {
            let first = ip.segments()[0];
            ip.is_unspecified()
                || ip.is_loopback()
                || (first & 0xfe00) == 0xfc00
                || (first & 0xffc0) == 0xfe80
                || (first & 0xff00) == 0xff00
                || ip
                    .to_ipv4_mapped()
                    .is_some_and(|ip| is_private_or_internal_ip(IpAddr::V4(ip)))
        }
    }
}

pub fn validate_fetch_url(raw_url: &str, allow_private: bool) -> Result<Url, FetchUrlPolicyError> {
    let url = Url::parse(raw_url).map_err(|_| FetchUrlPolicyError::InvalidUrl)?;
    if !matches!(url.scheme(), "http" | "https") {
        return Err(FetchUrlPolicyError::UnsupportedProtocol(
            url.scheme().to_owned(),
        ));
    }
    if !url.username().is_empty() || url.password().is_some() {
        return Err(FetchUrlPolicyError::EmbeddedCredentials);
    }
    let host = url.host_str().ok_or(FetchUrlPolicyError::InvalidUrl)?;
    if !allow_private && is_private_hostname(host) {
        return Err(FetchUrlPolicyError::PrivateAddress(host.to_owned()));
    }
    Ok(url)
}

pub async fn resolve_fetch_target(
    raw_url: &str,
    allow_private: bool,
    resolver: &dyn AgentDnsResolver,
) -> Result<ResolvedWebTarget, FetchUrlPolicyError> {
    let url = validate_fetch_url(raw_url, allow_private)?;
    let host = url.host_str().ok_or(FetchUrlPolicyError::InvalidUrl)?;
    let port = url
        .port_or_known_default()
        .ok_or(FetchUrlPolicyError::InvalidUrl)?;
    let resolved_ips = match host.parse::<IpAddr>() {
        Ok(ip) => vec![ip],
        Err(_) => resolver.resolve(host, port).await?,
    };
    if resolved_ips.is_empty() {
        return Err(FetchUrlPolicyError::EmptyDnsResult);
    }
    if !allow_private {
        if let Some(ip) = resolved_ips
            .iter()
            .copied()
            .find(|ip| is_private_or_internal_ip(*ip))
        {
            return Err(FetchUrlPolicyError::PrivateAddress(ip.to_string()));
        }
    }
    Ok(ResolvedWebTarget { url, resolved_ips })
}

#[derive(Clone)]
pub struct AgentWebTools {
    resolver: Arc<dyn AgentDnsResolver>,
    transport: Arc<dyn AgentWebHttpTransport>,
    search: Option<Arc<dyn AgentWebSearch>>,
    allow_private_fetch: bool,
}

impl AgentWebTools {
    pub fn new(
        resolver: Arc<dyn AgentDnsResolver>,
        transport: Arc<dyn AgentWebHttpTransport>,
        search: Option<Arc<dyn AgentWebSearch>>,
        allow_private_fetch: bool,
    ) -> Self {
        Self {
            resolver,
            transport,
            search,
            allow_private_fetch,
        }
    }

    pub fn has_search(&self) -> bool {
        self.search.is_some()
    }

    pub async fn search_is_configured(&self) -> Result<bool, AgentPortError> {
        match &self.search {
            Some(search) => search.is_configured().await,
            None => Ok(false),
        }
    }

    pub async fn search(&self, query: &str) -> String {
        let Some(search) = &self.search else {
            return "Web search is not configured.".to_owned();
        };
        match search.search(query, WEB_SEARCH_MAX_RESULTS).await {
            Ok(results) => {
                let serialized = serde_json::to_string(&results_to_json(&results))
                    .unwrap_or_else(|_| "[]".to_owned());
                truncate_utf8(&serialized, WEB_SEARCH_RESULT_MAX_BYTES)
            }
            Err(error) => format!(
                "Web search failed: {}",
                redact_secrets(error.message()).text
            ),
        }
    }

    pub async fn fetch_url(&self, raw_url: &str) -> String {
        let mut current_url = raw_url.to_owned();
        for _ in 0..=FETCH_URL_MAX_REDIRECTS {
            let target = match resolve_fetch_target(
                &current_url,
                self.allow_private_fetch,
                self.resolver.as_ref(),
            )
            .await
            {
                Ok(target) => target,
                Err(error) => return error.to_string(),
            };
            let base_url = target.url.clone();
            let response = match self
                .transport
                .send(WebHttpRequest {
                    method: WebHttpMethod::Get,
                    target,
                    timeout: FETCH_URL_TIMEOUT,
                    max_body_bytes: FETCH_URL_MAX_BODY_BYTES,
                    headers: BTreeMap::from([(
                        "Accept".to_owned(),
                        "text/html,application/xhtml+xml,text/plain,*/*".to_owned(),
                    )]),
                    body: Vec::new(),
                })
                .await
            {
                Ok(response) => response,
                Err(error) => {
                    return format!("Fetch failed: {}", redact_secrets(error.message()).text);
                }
            };

            if (300..400).contains(&response.status) {
                let Some(location) = header(&response.headers, "location") else {
                    return format!(
                        "Fetch failed: HTTP {} redirect without Location header",
                        response.status
                    );
                };
                current_url = match base_url.join(location) {
                    Ok(url) => url.to_string(),
                    Err(_) => return "Fetch failed: invalid redirect URL".to_owned(),
                };
                continue;
            }
            if !(200..300).contains(&response.status) {
                return format!("Fetch failed: HTTP {}", response.status);
            }

            let body_len = response.body.len().min(FETCH_URL_MAX_BODY_BYTES);
            let body = String::from_utf8_lossy(&response.body[..body_len]);
            let content_type = header(&response.headers, "content-type").unwrap_or_default();
            let text = if content_type.contains("text/html")
                || content_type.contains("application/xhtml")
            {
                match extract_html_text(&body) {
                    Ok(text) => text,
                    Err(error) => return format!("Fetch failed: {}", error.message()),
                }
            } else {
                body.into_owned()
            };
            return wrap_untrusted(
                &truncate_utf8(&text, FETCH_URL_TEXT_MAX_BYTES),
                UntrustedContentKind::Web,
            );
        }
        format!("Fetch failed: too many redirects (>{FETCH_URL_MAX_REDIRECTS})")
    }
}

fn header<'a>(headers: &'a BTreeMap<String, String>, name: &str) -> Option<&'a str> {
    headers
        .iter()
        .find(|(key, _)| key.eq_ignore_ascii_case(name))
        .map(|(_, value)| value.as_str())
}

fn results_to_json(results: &[WebSearchResultItem]) -> Vec<serde_json::Value> {
    results
        .iter()
        .map(|item| {
            serde_json::json!({
                "title": item.title,
                "url": item.url,
                "snippet": item.snippet,
            })
        })
        .collect()
}

fn truncate_utf8(text: &str, max_bytes: usize) -> String {
    if text.len() <= max_bytes {
        return text.to_owned();
    }
    let mut end = max_bytes;
    while !text.is_char_boundary(end) {
        end -= 1;
    }
    format!("{}\n[truncated]", &text[..end])
}

struct HtmlPatterns {
    strip_blocks: Vec<Regex>,
    strip_tags: Regex,
    collapse_space: Regex,
    collapse_lines: Regex,
}

static HTML_PATTERNS: LazyLock<Result<HtmlPatterns, regex::Error>> = LazyLock::new(|| {
    let strip_blocks = [
        "script", "style", "noscript", "template", "svg", "nav", "header", "footer", "aside",
        "iframe",
    ]
    .into_iter()
    .map(|tag| Regex::new(&format!(r"(?is)<{tag}\b[^>]*>.*?</{tag}\s*>")))
    .collect::<Result<Vec<_>, _>>()?;
    Ok(HtmlPatterns {
        strip_blocks,
        strip_tags: Regex::new(r"(?s)<[^>]*>")?,
        collapse_space: Regex::new(r"[ \t\u{a0}]+")?,
        collapse_lines: Regex::new(r"(?:[ \t]*\n[ \t]*){2,}")?,
    })
});

fn extract_html_text(html: &str) -> Result<String, AgentPortError> {
    let patterns = HTML_PATTERNS
        .as_ref()
        .map_err(|_| AgentPortError::new("HTML sanitizer initialization failed"))?;
    let mut value = html.to_owned();
    for pattern in &patterns.strip_blocks {
        value = pattern.replace_all(&value, " ").into_owned();
    }
    value = patterns.strip_tags.replace_all(&value, " ").into_owned();
    for (entity, replacement) in [
        ("&nbsp;", " "),
        ("&amp;", "&"),
        ("&lt;", "<"),
        ("&gt;", ">"),
        ("&quot;", "\""),
        ("&#39;", "'"),
    ] {
        value = value.replace(entity, replacement);
    }
    value = patterns
        .collapse_space
        .replace_all(&value, " ")
        .into_owned();
    Ok(patterns
        .collapse_lines
        .replace_all(value.trim(), "\n\n")
        .into_owned())
}

#[cfg(test)]
mod tests {
    use std::sync::Mutex;

    use super::*;

    struct FixedResolver {
        answers: Mutex<Vec<Vec<IpAddr>>>,
    }

    #[async_trait]
    impl AgentDnsResolver for FixedResolver {
        async fn resolve(&self, _host: &str, _port: u16) -> Result<Vec<IpAddr>, AgentPortError> {
            Ok(self.answers.lock().expect("DNS answers").remove(0))
        }
    }

    #[test]
    fn blocks_literal_private_and_obfuscated_numeric_hosts() {
        for host in [
            "localhost",
            "127.0.0.1",
            "10.0.0.1",
            "169.254.169.254",
            "172.16.0.1",
            "192.168.1.1",
            "::1",
            "fc00::1",
            "fe80::1",
            "2130706433",
            "127.1",
            "0177.0.0.1",
            "0x7f000001",
        ] {
            assert!(is_private_hostname(host), "expected private: {host}");
        }
        assert!(!is_private_hostname("example.com"));
        assert!(!is_private_hostname("8.8.8.8"));
    }

    #[tokio::test]
    async fn rechecks_every_dns_answer_and_rejects_mixed_private_results() {
        let resolver = FixedResolver {
            answers: Mutex::new(vec![vec![
                "93.184.216.34".parse().expect("public IP"),
                "127.0.0.1".parse().expect("private IP"),
            ]]),
        };
        let error = resolve_fetch_target("https://example.com/path", false, &resolver)
            .await
            .expect_err("mixed DNS answer must be rejected");
        assert!(matches!(error, FetchUrlPolicyError::PrivateAddress(_)));
    }

    #[test]
    fn request_debug_omits_path_query_and_headers() {
        let request = WebHttpRequest {
            method: WebHttpMethod::Get,
            target: ResolvedWebTarget {
                url: Url::parse("https://example.com/private?token=sk-abcdefghijklmnop")
                    .expect("URL"),
                resolved_ips: vec!["93.184.216.34".parse().expect("IP")],
            },
            timeout: FETCH_URL_TIMEOUT,
            max_body_bytes: 10,
            headers: BTreeMap::from([(
                "Authorization".to_owned(),
                "Bearer sk-abcdefghijklmnop".to_owned(),
            )]),
            body: b"body-secret-marker".to_vec(),
        };
        let debug = format!("{request:?}");
        assert!(!debug.contains("private"));
        assert!(!debug.contains("token="));
        assert!(!debug.contains("Bearer"));
        assert!(!debug.contains("sk-"));
        assert!(!debug.contains("body-secret-marker"));
    }
}
